src/onefetch/repo.rs

```diff
@@ -0,0 +1,168 @@
+    repo: Repository,
+}
+
+impl Repo {
+    pub fn new(repo_path: &str) -> Result<Self> {
+        let repo = Repository::discover(repo_path)?;
+        Ok(Self { repo })
+    }
+    pub fn get_work_dir(&self) -> Result<String> {
+        if let Some(workdir) = self.work_dir()?.to_str() {
+            Ok(workdir.to_string())
+        } else {
+            Err("invalid workdir".into())
+        }
+    }
+
+    pub fn get_number_of_tags(&self) -> Result<usize> {
+        Ok(self.repo.tag_names(None)?.len())
+    }
+
+    pub fn get_number_of_branches(&self) -> Result<usize> {
+        Ok(self.repo.branches(Some(BranchType::Remote))?.count() - 1)
+    }
+
+    pub fn get_git_username(&self) -> Result<String> {
+        let config = self.repo.config()?;
+        let username = config.get_entry("user.name")?;
+        Ok(username.value().unwrap_or("unknown").to_string())
+    }
+
+    pub fn get_version(&self) -> Result<String> {
+        let mut version_name = String::new();
+        let mut most_recent: i64 = 0;
+
+        self.repo.tag_foreach(|id, name| {
+            if let Ok(name) = String::from_utf8(name[10..name.len()].into()) {
+                let mut current_time: i64 = 0;
+                if let Ok(tag) = self.repo.find_tag(id) {
+                    if let Ok(c) = self.repo.find_commit(tag.target_id()) {
+                        current_time = c.time().seconds();
+                    }
+                } else if let Ok(c) = self.repo.find_commit(id) {
+                    current_time = c.time().seconds();
+                }
+                if current_time > most_recent {
+                    most_recent = current_time;
+                    version_name = name;
+                }
+
+                return true;
+            }
+            false
+        })?;
+
+        Ok(version_name)
+    }
+
+    pub fn get_pending_changes(&self) -> Result<String> {
+        let statuses = self.repo.statuses(Some(
+            StatusOptions::default()
+                .show(StatusShow::Workdir)
+                .update_index(true)
+                .include_untracked(true)
+                .renames_head_to_index(true)
+                .recurse_untracked_dirs(true),
+        ))?;
+
+        let mut deleted: usize = 0;
+        let mut added: usize = 0;
+        let mut modified: usize = 0;
+
+        for e in statuses.iter() {
+            let s: Status = e.status();
+            if s.is_index_new() || s.is_wt_new() {
+                added += 1;
+            } else if s.is_index_deleted() || s.is_wt_deleted() {
+                deleted += 1;
+            } else {
+                modified += 1;
+            }
+        }
+
+        let mut result = String::new();
+        if modified > 0 {
+            result = format!("{}+-", modified)
+        }
+
+        if added > 0 {
+            result = format!("{} {}+", result, added);
+        }
+
+        if deleted > 0 {
+            result = format!("{} {}-", result, deleted);
+        }
```

Comment:
Is it possible for at least two of modifed, added, and deleted to have a value > 0? If `added > 0 && deleted > 0`, won't this cause the result to only include the deleted lines?

Reply:
I'm not sure to get it...at each step we **append** to the previous value of `result` either added or deleted.

Reply:
This might be behavior that's currently in `master`, too.

```bash
# in onefetch repo @ master
mkdir test-repo
cd test-repo
git init
echo print 1 > test.py
echo print 2 >> test.py
echo print 3 >> test.py
echo print 4 >> test.py
echo print 5 >> test.py
git add --all
git commit -m "Init"
echo print 0 > test.py
echo print 1 >> test.py
echo print 3  >> test.py
echo print 4 >> test.py
echo print 51 >> test.py
cd -
cargo run -- test-repo
```

Onefetch will report **Pending**: 1+-
(one modification)

however,
```bash
cd test-repo
git diff
```
Reports
```diff
+print 0
 print 1
-print 2
 print 3
 print 4
-print 5
+print 51
```
Which is 1 addition, 1 deletion, and 1 modification.

Does that make sense? Am I interpreting how Onefetch is supposed to display pending changes properly?

Reply:
Oh, nevermind, I see that it's a count of pending *file* changes, not line changes :facepalm:
Please ignore :laughing: